File: mod-common/src/main/java/com/dreamdisplays/net/VersionPacket.java
package com.dreamdisplays.net;

import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.resources.ResourceLocation;
import com.dreamdisplays.PlatformlessInitializer;

// Packet for sending mod version information
public record VersionPacket(String version) implements CustomPacketPayload {
    public static final Type<VersionPacket> PACKET_ID =
            new Type<>(ResourceLocation.fromNamespaceAndPath(PlatformlessInitializer.MOD_ID, "version"));

    public static final StreamCodec<FriendlyByteBuf, VersionPacket> PACKET_CODEC =
            StreamCodec.of(
                    (buf, packet) -> ByteBufCodecs.STRING_UTF8.encode(buf, packet.version()),
                    (buf) -> {
                        String version = ByteBufCodecs.STRING_UTF8.decode(buf);
                        return new VersionPacket(version);
                    });

    @Override
    public Type<? extends CustomPacketPayload> type() {
        return PACKET_ID;
    }
}

File: common/src/main/kotlin/com/dreamdisplays/net/common/VersionPacket.kt
package com.dreamdisplays.net.common

import com.dreamdisplays.net.common.helpers.createType
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.codec.StreamCodec.of
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.minecraft.network.protocol.common.custom.CustomPacketPayload.Type

data class VersionPacket(val version: String) : CustomPacketPayload {
    override fun type(): Type<out CustomPacketPayload> = PACKET_ID

    companion object {
        val PACKET_ID: Type<VersionPacket> = createType("version")
        val PACKET_CODEC: StreamCodec<FriendlyByteBuf, VersionPacket> = of(
            { buf, packet -> buf.writeUtf(packet.version) },
            { buf -> VersionPacket(buf.readUtf()) }
        )
    }
}
